File: Platforms/Windows/Reign.Video.D3D9/IndexBuffer.cpp
#include "pch.h"
#include "IndexBuffer.h"

namespace Reign
{namespace Video
{namespace D3D9
{
	#pragma region Constructors
	IndexBuffer::IndexBuffer(DisposableI^ parent, BufferUsages bufferUsage)
	: IndexBufferI(parent, bufferUsage)
	{
		init(parent, nullptr);
	}

	IndexBuffer::IndexBuffer(DisposableI^ parent, BufferUsages bufferUsage, array<int>^ indices)
	: IndexBufferI(parent, bufferUsage)
	{
		init(parent, indices);
	}

	IndexBuffer::IndexBuffer(DisposableI^ parent, BufferUsages bufferUsage, array<int>^ indices, bool _32BitIndices)
	: IndexBufferI(parent, bufferUsage, _32BitIndices)
	{
		init(parent, indices);
	}

	void IndexBuffer::init(DisposableI^ parent, array<int>^ indices)
	{
		null();
		try
		{
			video = parent->FindParentOrSelfWithException<Video^>();
			if (indices) Init(indices);
		}
		catch (Exception^ ex)
		{
			delete this;
			throw ex;
		}
	}

	IndexBuffer::~IndexBuffer()
	{
		disposeChilderen();
		if (indexBuffer) indexBuffer->Release();
		null();
	}

	void IndexBuffer::null()
	{
		indexBuffer = 0;
		video = nullptr;
	}
	#pragma endregion

	#pragma region Methods
	void IndexBuffer::Init(array<int>^ indices)
	{
		IndexBufferI::Init(indices);
		if (indexBuffer)
		{
			indexBuffer->Release();
			indexBuffer = 0;
		}

		uint bufferSize = indexCount * indexByteSize;
		IDirect3DIndexBuffer9* indexBufferTEMP;
		if (FAILED(video->Device->CreateIndexBuffer(bufferSize, D3DUSAGE_WRITEONLY, _32BitIndices ? D3DFMT_INDEX32 : D3DFMT_INDEX16, video->IsExDevice ? D3DPOOL_DEFAULT : D3DPOOL_MANAGED, &indexBufferTEMP, 0)))
		{
			Debug::ThrowError(L"IndexBuffer", L"Failed to create IndexBuffer");
		}
		indexBuffer = indexBufferTEMP;
		Update(indices, indexCount);
	}

	void IndexBuffer::Update(array<int>^ indices, int updateCount)
	{
		void *data = 0;
		uint bufferSize = indexByteSize * updateCount;
		indexBuffer->Lock(0, bufferSize, (void**)&data, 0);

		void* indicesPtr = 0;
		short* _16BitIndices = 0;
		if (_32BitIndices)
		{
			pin_ptr<int>indicesTEMP = &indices[0];
			indicesPtr = indicesTEMP;
		}
		else
		{
			_16BitIndices = new short[updateCount];
			for (int i = 0; i != updateCount; ++i)
			{
				_16BitIndices[i] = indices[i];
			}
			indicesPtr = _16BitIndices;
		}
		
		memcpy(data, indicesPtr, bufferSize);
		indexBuffer->Unlock();
		if (_16BitIndices) delete _16BitIndices;
	}

	void IndexBuffer::enable()
	{
		video->Device->SetIndices(indexBuffer);
	}
	#pragma endregion
}
}
}

File: Platforms/Win32/Reign.Video.D3D9/Buffers/IndexBuffer.cs
using System;
using Reign_Video_D3D9_Component;
using Reign.Video;
using Reign.Core;

namespace Reign.Video.D3D9
{
	public class IndexBuffer : IndexBufferI
	{
		#region Properties
		internal IndexBufferCom com;
		#endregion

		#region Constructors
		public static IndexBuffer New(DisposableI parent, BufferUsages usage, int[] indices)
		{
			return new IndexBuffer(parent, usage, indices);
		}

		public IndexBuffer(DisposableI parent, BufferUsages usage, int[] indices)
		: base(parent, usage, indices)
		{
			try
			{
				var video = parent.FindParentOrSelfWithException<Video>();
				com = new IndexBufferCom(video.com);
				var error = com.Init(indices, indexCount, indexByteSize, REIGN_D3DUSAGE.WRITEONLY, _32BitIndices);
				if (error == IndexBufferErrors.IndexBuffer) Debug.ThrowError("IndexBuffer", "Failed to create IndexBuffer");
			}
			catch (Exception e)
			{
				Dispose();
				throw e;
			}
		}

		public override void Dispose()
		{
			disposeChilderen();
			if (com != null)
			{
				com.Dispose();
				com = null;
			}
			base.Dispose();
		}
		#endregion

		#region Methods
		public override void Update(int[] indices, int updateCount)
		{
			com.Update(indices, updateCount, indexByteSize, _32BitIndices);
		}
		#endregion
	}
}
